hail/src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -62,184 +65,322 @@ case class IndexMetadata(
+  val eltType: PCanonicalStruct = types.coerce[PCanonicalStruct](arrayType.elementType)
+  private val elt = new PCanonicalBaseStructSettable(eltType, sb.newSettable[Long](s"${name}_elt_off"))
 
-  private def writeInternalNode(node: InternalNodeBuilder, level: Int, isRoot: Boolean = false): Long = {
-    val indexFileOffset = trackedOS.bytesWritten
+  def length: Code[Int] = len
 
-    val info = if (node.size > 0) {
-      val firstChild = node.getChild(0)
-      val firstIndex = firstChild.firstIndex
-      val firstKey = firstChild.firstKey
-      val firstRecordOffset = firstChild.firstRecordOffset
-      val firstAnnotation = firstChild.firstAnnotation
-      IndexNodeInfo(indexFileOffset, firstIndex, firstKey, firstRecordOffset, firstAnnotation)
-    } else {
-      assert(isRoot && level == 0)
-      null
-    }
+  def loadFrom(cb: EmitCodeBuilder, a: Code[Long], l: Code[Int]): Unit = {
+    cb += (aoff := a)
+    cb += (len := l)
+  }
 
-    internalEncoder.writeByte(1)
+  def create(cb: EmitCodeBuilder, dest: Code[Long]): Unit = {
+    cb += (aoff := arrayType.allocate(region, maxSize))
+    cb += arrayType.stagedInitialize(aoff, maxSize)
+    cb += PCode(arrayType, aoff).store(cb.emb, region, dest)
+    cb += (len := 0)
+  }
 
-    val regionOffset = node.write(rvb)
-    internalEncoder.writeRegionValue(regionOffset)
-    internalEncoder.flush()
+  def storeLength(cb: EmitCodeBuilder): Unit = cb += arrayType.storeLength(aoff, length)
 
-    region.clear()
-    node.clear()
+  def setFieldValue(cb: EmitCodeBuilder, name: String, field: PCode): Unit = {
+    cb += eltType.setFieldPresent(elt.a, name)
+    cb += StagedRegionValueBuilder.deepCopy(cb.emb.ecb, region, eltType.fieldType(name), field.code, eltType.fieldOffset(elt.a, name))
+  }
 
-    if (!isRoot) {
-      if (level + 1 == internalNodeBuilders.length)
-        internalNodeBuilders += new InternalNodeBuilder(keyType, annotationType)
-      val parent = internalNodeBuilders(level + 1)
-      if (parent.size == branchingFactor)
-        writeInternalNode(parent, level + 1)
-      parent += info
-    }
+  def setField(cb: EmitCodeBuilder, name: String, v: => IEmitCode): Unit =
+    v.consume(cb,
+      cb += eltType.setFieldMissing(elt.a, name),
+      setFieldValue(cb, name, _))
 
-    indexFileOffset
+  def addChild(cb: EmitCodeBuilder): Unit = {
+    loadChild(cb, len)
+    cb += (len := len + 1)
   }
+  def loadChild(cb: EmitCodeBuilder, idx: Code[Int]): Unit = cb += elt.store(PCode(eltType, arrayType.elementOffset(aoff, idx)))
+  def getLoadedChild: PBaseStructValue = elt
 
-  private def writeLeafNode(): Long = {
-    val indexFileOffset = trackedOS.bytesWritten
+  def getChild(idx: Value[Int]): PCode = PCode(eltType, arrayType.elementOffset(aoff, idx))
+}
 
-    assert(leafNodeBuilder.size > 0)
+class StagedIndexWriterUtils(ib: Settable[IndexWriterUtils]) {
+  def create(path: Code[String], fs: Code[FS], meta: Code[StagedIndexMetadata]): Code[Unit] =
```

Comment:
We should make these `Code[Unit]` methods take a `CodeBuilder` and return `Unit`.